R/utils.R

```diff
@@ -329,3 +329,381 @@ post_x <- function(x, params) {
+  }
+
+  relevant_leaves <- updates_relevant_leaves$relevant_leaves[,`:=` (f_idx = .I, f_idx_uncond = f_idx)][]
+  
+  cnt_new <- setcolorder(merge(relevant_leaves, updates_relevant_leaves$cnt_new, by.x = c("c_idx", "f_idx_uncond"), by.y = c("c_idx", "f_idx"), sort = F), c("f_idx","c_idx","variable","min","max","val","cvg_factor"))[]
+  cat_new <- setcolorder(merge(relevant_leaves, updates_relevant_leaves$cat_new, by.x = c("c_idx", "f_idx_uncond"), by.y = c("c_idx", "f_idx"), sort = F), c("f_idx","c_idx","variable","val","prob","cvg_factor"))[]
+
+  if(relevant_leaves[,uniqueN(c_idx)] < nconds_conditioned) {
+    if(relevant_leaves[,uniqueN(c_idx)] == 0 & row_mode == "or") {
+      stop("For all entered evidence rows, no matching leaves could be found. This is probably because evidence lies outside of the distribution calculated by FORDE. For continuous data, consider setting epsilon>0 or finite_bounds=FALSE in forde(). For categorical data, consider setting alpha>0 in forde()")
+    } else {
+      warning("For some entered evidence rows, no matching leaves could be found. This is probably because evidence lies outside of the distribution calculated by FORDE. For continuous data, consider setting epsilon>0 or finite_bounds=FALSE in forde(). For categorical data, consider setting alpha>0 in forde()")
+      conds_impossible <- conds_conditioned[!(conds_conditioned %in% relevant_leaves[,unique(c_idx)])]
+      relevant_leaves <- setorder(rbind(relevant_leaves, data.table(c_idx = conds_impossible, f_idx = NA_integer_, f_idx_uncond = NA_integer_)))
+    }
+  }
+  
+  forest_new <- merge(relevant_leaves, forest, by.x = "f_idx_uncond", by.y = "f_idx", all.x = T, sort = F)
+  setnames(forest_new,"cvg","cvg_arf")
+  
+  cvg_new <- rbind(cat_new[,.(f_idx, c_idx, cvg_factor)], cnt_new[, .(f_idx, c_idx, cvg_factor)])
+  unique(cvg_new[,.(c_idx, f_idx)])
+
+  if(nrow(cvg_new) > 0) {
+    cvg_new[,cvg_factor := log(cvg_factor)]
+    cvg_new <- cvg_new[, .(cvg_factor = sum(cvg_factor)), keyby = f_idx]
+    cvg_new <- cbind(cvg_new, forest_new[, .(c_idx,cvg_arf = log(cvg_arf))])
+    cvg_new[,`:=` (cvg = cvg_factor + cvg_arf, cvg_factor = NULL, cvg_arf = NULL)]
+    
+    if(row_mode == "or") {
+      if(cvg_new[,all(cvg == - Inf)]) {
+        warning("All leaves have zero likelihood. This is probably because evidence contains an (almost) impossible combination.")
+        cvg_new[, cvg := 1/.N]
+      } else {
+        cvg_new[, cvg := exp(cvg - max(cvg))]
+        cvg_new <- cvg_new[cvg > 0,][,cvg := cvg / sum(cvg)]
+      }
+    } else {
+      cvg_new[, leaf_zero_lik := all(cvg == -Inf), by = c_idx]
+      if(any(cvg_new[, leaf_zero_lik])) {
+        warning("All leaves have zero likelihood for some entered evidence rows. This is probably because evidence contains an (almost) impossible combination.")
+        cvg_new[leaf_zero_lik == T, cvg := 1/.N, by = c_idx]
+      } else {
```

Comment:
Shouldn't this part (in else) also be run, even with there are leaves with zero likelihood? 
Now if any leaf has zero likelihood, the scaling is not done for any of the leaves.